### PR TITLE
Add debugmsg capture for tests

### PR DESCRIPTION
**Pull request by akirashirosawa**
#### Summary
SUMMARY: None
#### Purpose of change
For the most complete testing, including edge cases, necessary to `debugmsg` did not automatically fail tests results. If the test suggests `debugmsg` output, need to be able to handle it. At the moment, tests related to `item::ammo_set` call `debugmsg` (`"set 308 ammo in the 9mm gun with magazine"`, `"set 9mm ammo in the small box"`, etc). In the future, when adding `debugmsg` to the `item::put_in`(PR #41677), tests associated with it will also fail (`"wrong size battery for flashlight"`, `"it cannot hold an item 15 cm in length"`).

We can just remove tests that assume `debugmsg`, and/or don't add `debugmsg` to the `item::put_in`. But adding `debugmsg` to the`item::put_in` allowed finding and fixing a huge number of bugs ( > 10, see #41677, #43149). And test `"set 9mm ammo in the small box"` allowed me to catch a bug while fixes `item::ammo_set`. So I think it's worth it.
#### Describe the solution
The idea is to make a function call through a pointer and return the result, messages, and at the same time prevent the `error_observed` flag from being set.
```cpp
// capturing function
std::string capture_debugmsg_during( const std::function<void()> &func )
{
    capture_debugmsg capture;
    func();
    return capture.dmsg();
}

// usage in tests
WHEN( "set 9mm ammo in the small box" ) {
    std::string dmsg = capture_debugmsg_during( [&box, &ammo9mm_id]() {
        box.ammo_set( ammo9mm_id, 10 );
    } );
    THEN( "get debugmsg with \"Tried to set invalid ammo of 9mm for box_small\"" ) {
        CHECK_THAT( dmsg, Catch::EndsWith( "Tried to set invalid ammo of 9mm for box_small" ) );
        AND_THEN( "small box still empty" ) {
            CHECK( box.is_container_empty() );
        }
    }
```
#### Describe alternatives you've considered
1. May add the `capture_debugmsg` class to the header file so that can use capture  not wrapeed in the `capture_debugmsg_during`, but I find it redundant.
2. Despite capturing, output debug messages and stacktrace to stdout. I think this will spam successful tests and make the output sloppy. If something goes wrong, can always remove the capture to get the stacktrace.
#### Testing
I have uncommented the tests calling `debugmsg`, all tests passed successfully.
#### Additional context
After PR #41677 will be merged, I'll also add `debugmsg` capturing for `item::put_in` tests (where the TODO comments is now).

### CODE REVIEW

File: src/debug.h [hunk 1]
```diff
@@ -197,6 +197,50 @@ void limitDebugClass( int );
...
+capture_ret<R> CAPTURE_DEBUGMSG_S( F func, Args... args )
+{
+    capture_ret<R> ret;
+    captureStart();
+    ret.result = func( args... );
+    ret.dmsg = captureEnd();
+    ret.error_observed = !ret.dmsg.empty();
+    return ret;
+}
+
```
Comment (Qrox) [inline src/debug.h:243]: Since these are only supposed to be used in the tests, you can add a header and a cpp file to the test code that contains `captureStart`, `captureEnd`, but not `CAPTURE_DEBUGMSG` or `CAPTURE_DEBUGMSG_S`. Instead return `dmsg` and `error_observed` directly from `captureEnd`, so you can execute whatever code before calling `captureEnd` instead of being limited to a member or non-member function call and a single templated return value. To access `capturing` and `captured` in both cpp files, you can define them in `debug.cpp` without the `static` keyword, and declare them in the new cpp file using the `extern` keyword (`extern bool capturing;`)

You can also replace `captureStart` and `captureEnd` with a `captureDebugmsg` object that calls `captureStart` on initialization and calls `captureEnd` when explicitly asked to return the captured debug message and when the object is deconstructed. This ensures that `captureEnd` is always called appropriately without creating a function like `CAPTURE_DEBUGMSG`.

**Comment by akirashirosawa** [comment]
@Qrox 
Initially, I used `captureStart` and `captureEnd` right in tests. I did not like it, because I wanted to ensure that `captureEnd` was called, so I turned to `CAPTURE_DEBUGMSG` and templates hell.

It's not a bad idea to call `captureStart` on class constructor and `captureEnd` on destructor, but... Now, I have a better idea.

Thank you for statement `you can execute whatever code ... instead of being limited`. I wanted to answer you that there is no limitation, because you can call any code, even a million functions by simply wrapping them in a lambda. Like this:
```cpp
WHEN( "set 9mm ammo in the small box" ) {
    auto ret = CAPTURE_DEBUGMSG_S<bool>( [&box, &ammo9mm_id]() {
        box.ammo_set( ammo9mm_id, 10 );
        THEN( "small box still empty" ) {
            REQUIRE_FALSE( box.is_gun() );
            REQUIRE_FALSE( box.is_magazine() );
            CHECK( box.is_container_empty() );
        }
        return true; // result really no need here
    } );
    REQUIRE( ret.error_observed );
    CHECK_THAT( ret.dmsg, Catch::EndsWith( "Tried to set invalid ammo of 9mm for box_small" ) );
}
```
And then I realized that templates are not needed, when there is a lambda! :) Now I will rewrite everything.

**Comment by Qrox** [comment]
Well, if using less template parameters is your goal, then my suggested change does exactly that (it requires zero template parameters).

On the other hand, I've got another concerns about stacktrace: before this change stacktrace is printed whenever a debug message is shown, but now you skip printing the stacktrace and append only the debug message to the captured message, so it can be harder to find what exactly caused these debug messages to show up. Maybe the debug message should be printed as usual, just suppress setting the `error_observed` flag?

**Comment by akirashirosawa** [comment]
>Well, if using less template parameters is your goal

My goals is: 1. don't put `captureStart`/`captureEnd` to the `.h` file and 2. use only one function to capture.
And they both have been achieved:

```cpp
// debug.h
class captureDebugmsg
{
    public:
        captureDebugmsg();
        std::string dmsg();
        ~captureDebugmsg();
};
template<typename Lambda>
std::string captureDebugmsgLambda( Lambda func )
{
    captureDebugmsg capture;
    func();
    return capture.dmsg();
}
// debug.cpp
captureDebugmsg::captureDebugmsg()
{
    capturing = true;
    captured = "";
}
std::string captureDebugmsg::dmsg()
{
    capturing = false;
    return captured;
}
captureDebugmsg::~captureDebugmsg()
{
    capturing = false;
}
// *_test.cpp
// use lambda
std::string dmsg = captureDebugmsgLambda( [&box, &ammo9mm_id]() {
    box.ammo_set( ammo9mm_id, 10 );
} );
THEN( "get debugmsgw with \"Tried to set invalid ammo of 9mm for box_small\"" ) {
    REQUIRE( !dmsg.empty() );
    CHECK_THAT( dmsg, Catch::EndsWith( "Tried to set invalid ammo of 9mm for box_small" ) );
    // ...

// use class directly
captureDebugmsg capture;
cz75.ammo_set( ammo308_id, 2 );
std::string dmsg = capture.dmsg();
THEN( "get debugmsgw with \"Tried to set invalid ammo of 308 for cz75\"" ) {
    REQUIRE( !dmsg.empty() );
    CHECK_THAT( dmsg, Catch::EndsWith( "Tried to set invalid ammo of 308 for cz75" ) );
    // ...
```
As you suggested, I used the `captureDebugmsg` class. Because otherwise I would not be able not to raise `captureStart`/`captureEnd` in the `.h` file, because even when using lambda, templates are required to pass it as a parameter, and because templates must be released in the `.h` file (including `.cpp` in tests is not an option).

I will push the changes after build and test. When change any `.h` file, the project is rebuilt from zero, which takes about an hour :(

> Maybe the debug message should be printed as usual, just suppress setting the error_observed flag?

But then the stacktrace will spam the successful test results and they will look sloppy. The capture should only be used when you expect a specific `debugmsg`. If an `debugmsg` occurs that you did not expect, you can always simply remove the capture. As a compromise, maybe add `filename << ":" << line << "[" << funcname << "]"` to the message? So that you can immediately understand where the `debugmsg` came from. I'm not sure. Let other contributors give their opinion.

File: src/debug.cpp [hunk 2]
```diff
@@ -91,6 +91,28 @@ extern bool test_mode;
 /** Set to true when any error is logged. */
 static bool error_observed = false;
 
+/** If true, debug messages will be captured. */
```
Comment (jbytheway) [inline src/debug.cpp:94]: Maybe expand this comment to explain that this is used to test debugmsg calls in the unit tests.

File: src/debug.h [hunk 3]
```diff
@@ -197,6 +197,31 @@ void limitDebugClass( int );
...
+        std::string dmsg();
+        ~captureDebugmsg();
+};
+
+/**
+ * Capturing debugmsg for a lambda function
+ *  @return std::string debugmsg
+ */
+template<typename Lambda>
+std::string captureDebugmsgLambda( Lambda func )
```
Comment (jbytheway) [inline src/debug.h:218]:
```suggestion
std::string captureDebugmsgLambda( const std::function<void()> &func )
```
If you want this to not be a template, you can change it like this, and then move its declaration and (optionally) the definition of `class captureDebugmsg` into `debug.cpp`.

I'd also change the name, because it can be used with more than lambdas.  Maybe `captureDebugmsgDuring`?

File: src/debug.h [hunk 4]
```diff
@@ -197,6 +197,31 @@ void limitDebugClass( int );
...
+ * Class for capturing debugmsg.
+ * Сan also use a captureDebugmsgLambda function instead.
+ */
+class captureDebugmsg
+{
+    public:
+        captureDebugmsg();
+        // @return std::string debugmsg
+        std::string dmsg();
+        ~captureDebugmsg();
```
Comment (jbytheway) [inline src/debug.h:210]:
```suggestion
class capture_debugmsg
{
    public:
        capture_debugmsg();
        // @return std::string debugmsg
        std::string dmsg();
        ~capture_debugmsg();
```
We generally use `snake_case`.

File: tests/ammo_set_test.cpp [hunk 5]
```diff
@@ -74,14 +74,20 @@ TEST_CASE( "ammo_set", "[ammo_set][magazine][ammo]" )
...
-            //         CHECK( cz75mag_20rd.ammo_current().is_null() );
-            //     }
-            // }
+            WHEN( "set 308 ammo in the 9mm magazine" ) {
+                itype_id ammo308_id( "308" );
+                std::string dmsg = captureDebugmsgLambda( [&cz75mag_20rd, &ammo308_id]() {
+                    cz75mag_20rd.ammo_set( ammo308_id, 15 );
+                } );
+                THEN( "get debugmsg with \"Tried to set invalid ammo of 308 for cz75mag_20rd\"" ) {
+                    REQUIRE( !dmsg.empty() );
```
Comment (jbytheway) [inline tests/ammo_set_test.cpp:83]:
```suggestion
```
I think this is redundant.

File: tests/battery_mod_test.cpp [hunk 6]
```diff
@@ -342,6 +342,7 @@ TEST_CASE( "installing battery in tool", "[battery][tool][install]" )
 
         // Should fail to install the magazine
         REQUIRE( flashlight.contents.has_pocket_type( item_pocket::pocket_type::MAGAZINE_WELL ) );
+        // TODO capture debugmsg PR #41677 will be merged
```
Comment (jbytheway) [inline tests/battery_mod_test.cpp:345]: It's not clear what these TODO comments mean; can you expand them a little.

File: tests/ammo_set_test.cpp (hunk 5 above)
Comment (jbytheway) [reply tests/ammo_set_test.cpp:83]: (and similarly for the other test cases)

File: tests/battery_mod_test.cpp (hunk 6 above)
Comment (jbytheway) [reply tests/battery_mod_test.cpp:345]: Ah, I think I understand after re-reading your PR description.  Maybe change this to
```
// TODO capture and vefify debugmsg output once PR #41677 has been merged
```

File: src/debug.cpp (hunk 2 above)
Comment (akirashirosawa) [reply src/debug.cpp]: done

File: src/debug.h (hunk 3 above)
Comment (akirashirosawa) [reply src/debug.h]: done

File: src/debug.h (hunk 4 above)
Comment (akirashirosawa) [reply src/debug.h]: done

File: tests/battery_mod_test.cpp (hunk 6 above)
Comment (akirashirosawa) [reply tests/battery_mod_test.cpp]: I have already added another comment. In fact, this is TODO for me. I will do it either in this PR or PR #41677, whichever PR is the second. It was also added to trigger a merge conflict so I will not forget to resolve it.

File: tests/ammo_set_test.cpp (hunk 5 above)
Comment (akirashirosawa) [reply tests/ammo_set_test.cpp]: done

**Comment by akirashirosawa** [comment]
I still moved class `captureDebugmsg` (aka `capture_debugmsg`) to the `.cpp` file, as it seemed redundant to me. If someone in the future lacks the functionality of `capture_debugmsg_during`, he will always be able to raise it to a `.h` file.

**Comment by akirashirosawa** [comment]
@jbytheway Do you know how to speed up the build of the project after changing the `.h` file? Even if i just removed the extra whitespace in the header file, the project will be rebuilt for about an hour. Maybe i can make `clang` use more cores? It only uses one CPU core on my mac.

**Comment by ZhilkinSerg** [comment]
Are you using `ccache`?

**Comment by akirashirosawa** [comment]
> Are you using `ccache`?

I've never heard of this, so guess not)

**Comment by akirashirosawa** [comment]
It turns out that `doc/COMPILING/COMPILING.md` is worth reading :)
> If you have a multi-core computer you'd probably want to add -jX to the options, where X should roughly be twice the number of cores you have available.


> If you plan on keeping up with experimentals you should also install ccache, which will considerably speed-up partial builds.

Better late than never)

**Comment by akirashirosawa** [comment]
I did rebase and update PR description.
